File: RailFenceCipher.java
// pw7;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class RailFenceCipher {
    
    public static String encrypt(String plaintext, int depth) {
        if (depth < 2) {
            throw new IllegalArgumentException("Depth must be 2 or greater");
        }
        
        List<StringBuilder> rails = new ArrayList<>();
        for (int i = 0; i < depth; i++) {
            rails.add(new StringBuilder());
        }
        
        int currentRail = 0;
        boolean goingDown = true;
        
        for (char c : plaintext.toCharArray()) {
            rails.get(currentRail).append(c);
            
            if (goingDown) {
                currentRail++;
                if (currentRail == depth - 1) {
                    goingDown = false;
                }
            } else {
                currentRail--;
                if (currentRail == 0) {
                    goingDown = true;
                }
            }
        }
        
        StringBuilder ciphertext = new StringBuilder();
        for (StringBuilder rail : rails) {
            ciphertext.append(rail);
        }
        
        return ciphertext.toString();
    }
    
    public static String decrypt(String ciphertext, int depth) {
        if (depth < 2) {
            throw new IllegalArgumentException("Depth must be 2 or greater");
        }
        
        char[][] railMatrix = new char[depth][ciphertext.length()];
        for (int i = 0; i < depth; i++) {
            for (int j = 0; j < ciphertext.length(); j++) {
                railMatrix[i][j] = '\0';
            }
        }
        
        int currentRail = 0;
        boolean goingDown = true;
        
        for (int i = 0; i < ciphertext.length(); i++) {
            railMatrix[currentRail][i] = '*';
            
            if (goingDown) {
                currentRail++;
                if (currentRail == depth - 1) {
                    goingDown = false;
                }
            } else {
                currentRail--;
                if (currentRail == 0) {
                    goingDown = true;
                }
            }
        }
        
        int index = 0;
        for (int i = 0; i < depth; i++) {
            for (int j = 0; j < ciphertext.length(); j++) {
                if (railMatrix[i][j] == '*' && index < ciphertext.length()) {
                    railMatrix[i][j] = ciphertext.charAt(index++);
                }
            }
        }
        
        StringBuilder plaintext = new StringBuilder();
        currentRail = 0;
        goingDown = true;
        
        for (int i = 0; i < ciphertext.length(); i++) {
            plaintext.append(railMatrix[currentRail][i]);
            
            if (goingDown) {
                currentRail++;
                if (currentRail == depth - 1) {
                    goingDown = false;
                }
            } else {
                currentRail--;
                if (currentRail == 0) {
                    goingDown = true;
                }
            }
        }
        
        return plaintext.toString();
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Enter the depth (3 or higher): ");
        int depth = scanner.nextInt();
        scanner.nextLine();
        
        System.out.print("Enter the text to encrypt: ");
        String plaintext = scanner.nextLine();
        
        String encrypted = encrypt(plaintext, depth);
        System.out.println("Encrypted: " + encrypted);
        
        String decrypted = decrypt(encrypted, depth);
        System.out.println("Decrypted: " + decrypted);
    }
}

File: RailFenceCipher.py
#pw7

def encrypt(plaintext, depth):
    if depth < 2:
        raise ValueError("Depth must be 2 or greater")
    
    rails = [[] for _ in range(depth)]
    current_rail = 0
    going_down = True
    
    for char in plaintext:
        rails[current_rail].append(char)
        
        if going_down:
            current_rail += 1
            if current_rail == depth - 1:
                going_down = False
        else:
            current_rail -= 1
            if current_rail == 0:
                going_down = True
    
    ciphertext = ''.join([''.join(rail) for rail in rails])
    return ciphertext

def decrypt(ciphertext, depth):
    if depth < 2:
        raise ValueError("Depth must be 2 or greater")
    
    # Create a matrix to mark the rail positions
    rail_matrix = [['' for _ in range(len(ciphertext))] for _ in range(depth)]
    
    # Mark the positions with '*'
    current_rail = 0
    going_down = True
    
    for i in range(len(ciphertext)):
        rail_matrix[current_rail][i] = '*'
        
        if going_down:
            current_rail += 1
            if current_rail == depth - 1:
                going_down = False
        else:
            current_rail -= 1
            if current_rail == 0:
                going_down = True
    
    # Fill the marked positions with ciphertext characters
    index = 0
    for i in range(depth):
        for j in range(len(ciphertext)):
            if rail_matrix[i][j] == '*' and index < len(ciphertext):
                rail_matrix[i][j] = ciphertext[index]
                index += 1
    
    # Read the plaintext
    plaintext = []
    current_rail = 0
    going_down = True
    
    for i in range(len(ciphertext)):
        plaintext.append(rail_matrix[current_rail][i])
        
        if going_down:
            current_rail += 1
            if current_rail == depth - 1:
                going_down = False
        else:
            current_rail -= 1
            if current_rail == 0:
                going_down = True
    
    return ''.join(plaintext)

if __name__ == "__main__":
    depth = int(input("Enter the depth (3 or higher): "))
    plaintext = input("Enter the text to encrypt: ")
    
    encrypted = encrypt(plaintext, depth)
    print("Encrypted:", encrypted)
    
    decrypted = decrypt(encrypted, depth)
    print("Decrypted:", decrypted)